src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1715 @@
+			ships_heading_there.push_back(ship_and_plan.first.get(game));
+		}
+	}
+
+	// Find all shippingitems heading for the deleted dock.
+	// Those in warehouses can just be told to recalculate their route.
+	// Those on ships will be rerouted to whichever portdock the ships will visit next;
+	// they will be unloaded there and then recalculate their route.
+
+	for (PortDock* pd : fleet_.get_ports()) {
+		for (auto it = pd->waiting_.begin(); it != pd->waiting_.end();) {
+			if (it->destination_dock_.serial() == dock->serial()) {
+				sslog("found a shippingitem in port %u\n", pd->serial());
+				it->set_location(game, pd->warehouse_);
+				it->end_shipping(game);
+				it = pd->waiting_.erase(it);
+			} else {
+				++it;
+			}
+		}
+	}
+
+	for (Ship* ship : ships_heading_there) {
+		for (ShippingItem& si : ship->items_) {
+			if (si.destination_dock_.serial() == dock->serial()) {
+				sslog("found a shippingitem on %s\n", ship->get_shipname().c_str());
+				si.destination_dock_ = ship->get_destination();
+			}
+		}
+	}
+	sslog("--- port_removed maintenance complete ---\n\n");
+}
+
+void ShippingSchedule::ship_removed(const Game&, Ship* ship) {
+	auto it = plans_.find(ship);
+	assert(it != plans_.end());
+	plans_.erase(it);
+
+	auto i = last_actual_duration_recalculation_.find(ship);
+	if (i != last_actual_duration_recalculation_.end()) {
+		last_actual_duration_recalculation_.erase(i);
+	}
+
+	// Handling any items that were intended to be transported by this ship
+	// is deferred to the next call to update()
+}
+
+void ShippingSchedule::ship_added(Game& game, Ship& s) {
+	sslog("\nShippingSchedule::ship_added (%s)\n", s.get_shipname().c_str());
+	assert(!s.get_destination());
+	plans_[&s] = ShipPlan();
+	last_actual_duration_recalculation_[&s] = game.get_gametime();
+	if (fleet_.get_ports().empty()) {
+#ifndef NDEBUG
+		for (ShippingItem& si : s.items_) {
+			assert(!si.destination_dock_.is_set());
+		}
+#endif
+		sslog("No ports!\n\n");
+		return;
+	}
+
+	PortDock* closest = nullptr;
+	int32_t dist = 0;
+	for (PortDock* dock : fleet_.get_ports()) {
+		Path path;
+		int32_t d = -1;
+		s.calculate_sea_route(game, *dock, &path);
+		game.map().calc_cost(path, &d, nullptr);
+		assert(d >= 0);
+		if (!closest || d < dist) {
+			dist = d;
+			closest = dock;
+		}
+	}
+	assert(closest);
+	for (ShippingItem& si : s.items_) {
+		si.destination_dock_ = closest;
+	}
+	plans_[&s].push_back(SchedulingState(closest, false, dist));
+	s.set_destination(game, closest);
+	sslog("Sent to %u\n\n", closest->serial());
+}
+
+void ShippingSchedule::port_added(Game& game, PortDock& dock) {
+	sslog("\nShippingSchedule::port_added (%u)\n", dock.serial());
+	if (fleet_.count_ports() > 1) {
+		// nothing to do currently
+		sslog("nothing to do\n\n");
+		return;
+	}
+	// All ships are most likely panicking because they have
+	// no destination. Send them all to the new port.
+	for (Ship* ship : fleet_.get_ships()) {
```

Comment:
Mhh, I usually build a Shipwright next to a Port, If I have two such ports a the edges of my Mainland, a ship from one unfinished port  will trave to the one and only port, and may have to come back just a llittle later.

Maybe collect only a few, nearby ships? 

Juts a thought no need to change anything.

Reply:
To me this looks like a corner-case that only arises for the first ship if you build several ports at once. In the majority of cases it makes more sense to whisk any idle ships away to some port instantly IMHO.